### DISCUSSION TITLE
top-level awaits being recognised as Binops when source type is "unambiguous"

### DISCUSSION THREAD

**Discussion by meetesh06** [babel/babel#16776] [Q&A] [answered]
I am looking at this testcase: [test262](https://github.com/tc39/test262/blob/main/test/language/module-code/top-level-await/await-expr-regexp.js)

*I see that babel fails this test on the website; I don't know how the babel test pipeline works so I am not sure what to make of it! (would love to know how these tests apply to babel; as I am trying to create a s2s transformer aimed at transforming code to a subset of JS)*

**Source**
```javascript
// ...
await /x.y/g;
// ...
```

**Parsed AST**
```bash
BINOP -- Left: Identifier -- 'await' // <- Non module level code treats top level await as an identifier?? is it not a keyword??
             |
             |-- Right: Member Expression...
```

**Transformed Code**
```bash
let js3$5 = await; // <- ;(
let js3$6 = x.y;
let js3$4 = js3$5 / js3$6;
let js3$7 = g;
let js3$3 = js3$4 / js3$7;
```

Is it incorrect to assume that **top-level awaits imply module mode** for the parsed code, or are there cases when this assumption is incorrect?

**Comment by nicolo-ribaudo** [comment, marked as answer]
await is a valid identifier outside of modules, so as you discovered that file can be parsed in two different ways:
- as a script, equivalent to `(await / x.y) / g;`
- as a module with top-level await

**Reply by meetesh06** [reply]
Thanks 👍

**Comment by nicolo-ribaudo** [comment]
And yes, we run test262 as part of our CI :) You can find our results at https://github.com/babel/babel/actions/runs/10541220211/artifacts/1850744835
